Asking for Help lesson: Add How to Ask community guide (#25689)
TheOdinProject/curriculum: foundations/introduction/asking_for_help.md

Keep in mind that you’re actually supposed to figure out how to solve the assignments yourself, and figuring out an approach is essential to your learning journey. A much better question might look like this:

-   “Hey, I’m trying to return a string that displays the winner in Rock Paper Scissors, but I’m getting a syntax error on line 12. How can I fix this? Here’s my code.”

In sharing your attempt, people know what you’ve tried and won’t suggest things that might not have worked out for you. It also allows them to debug your current iteration of code, rather than sending you down a path that has you start all over again when you might be very close to a solution.

Now, if you’re completely stuck with where to start, it’s completely fine to let people know that you’re stuck. Asking where you can start and what you can research to get on the right track empowers you to be able to resolve issues largely on your own in the future, and might even empower you to help others with the same issue later on. It is also recommended that you share your pseudo code so people can nudge you in the right direction or correct any misunderstandings you may have.

> #### Don’t take asking for nore context to heart

Reviewer: ```suggestion
#### Don’t take asking for more context to heart
```

Author: Great catch! I must have done that it too much of a hurry!

### Tips for Getting the Best Help Possible

#### Always Provide Your Code and the Surrounding Context

> You can ask theoretical, vague questions, but you’re going to get theoretical, vague answers that probably won't be helpful to you and cause frustration on both ends as you and the person helping you dance around the real issue. The only time code doesn’t need to be provided is if you are asking a purely conceptual question outside of a specific project or code snippet. Additionally, provide as much context as is necessary, and zero in on the specific problem, such as pointing people to a specific function or line number in your code so people can give you the most relevant answers and quickly assist you.

Author: Thanks for the suggestion, but I do not believe the proposed wording is very clear. I am truly not a big fan of that paragraph, so I am going to give it a big re-write. What do you think of this?

```suggestion
When asking a question it is essential to provide your code, error message, terminal command, server output, etc. You should provide as much context as possible and zero in on the specific problem, such as pointing people to a specific function or line number in your code. You should try to make it as easy as possible for others to help you.

If your question does not provide any code or other relevant information, you will only receive theoretical and vague answers. This often causes frustration for everyone. You may become frustrated because the answers are not helpful. Those helping may become frustrated because as you add new information, the meaning of your original question changes and they wasted time trying to answer it. If you truly want to ask a purely conceptual question, you should indicate this in your question.
```

> You can ask theoretical, vague questions, but you’re going to get theoretical, vague answers that probably won't be helpful to you and cause frustration on both ends as you and the person helping you dance around the real issue. The only time code doesn’t need to be provided is if you are asking a purely conceptual question outside of a specific project or code snippet. Additionally, provide as much context as is necessary, and zero in on the specific problem, such as pointing people to a specific function or line number in your code so people can give you the most relevant answers and quickly assist you.

Reviewer 2: That's a good one too @rlmoser99

### Tips for Getting the Best Help Possible

> #### Always Provide Your Code and the Surrounding Context

Reviewer: ```suggestion
#### Always provide your code and the surrounding context
```

> When asking a question it is essential to provide your code, error message, terminal command, server output, etc. You should provide as much context as possible and zero in on the specific problem, such as pointing people to a specific function or line number in your code. You should try to make it as easy as possible for others to help you.

Reviewer 2: [Typo]
```suggestion
When asking a question, it is essential to provide your code, error message, terminal command, server output, etc. You should provide as much context as possible and zero in on the specific problem, such as pointing people to a specific function or line number in your code. You should also try to make it as easy as possible for others to help you.
```

Author: I'll add the comma. But I do not believe the "also" works here, but I do have a different wording in mind that this highlights.

> If your question does not provide any code or other relevant information, you will only receive theoretical and vague answers. This often causes frustration for everyone. You may become frustrated because the answers are not helpful. Those helping may become frustrated because as you add new information, the meaning of your original question changes and they wasted time trying to answer it. If you truly want to ask a purely conceptual question, you should indicate this in your question.

Reviewer: ```suggestion
If your question does not provide any code or other relevant information, this will often create unnecessary back and forth conversation to gather all the required information. If people are guessing what your problem is because they don't have all the details, their answers might be vague or technical, or not even the answer to your actual problem. This often causes frustration for everyone. You may become frustrated because the answers are not helpful. Those helping may become frustrated because as you add new information, the meaning of your original question changes and they wasted time trying to answer it. If you truly want to ask a purely conceptual question, you should indicate this in your question.
```

I reworded this a bit to add the "back and forth conversation". I often see this happening in our discord and they can sometimes make the path to an answer 20 messages long instead of the few it would have taken if the problem was clear from the start.

Author: I like the addition of the back and forth conversation, but the second sentence added a bit of repetition, since it also mentions that the answers are not helpful. Plus the back and forth conversation is the "this" that causes frustration. What do you think of this?

```suggestion
If your question does not provide any code or other relevant information, this will often create unnecessary back and forth conversation to gather all the required information. This often causes frustration for everyone. You may become frustrated because the answers are not helpful. Those helping may become frustrated because as you add new information, the meaning of your original question changes and they wasted time trying to answer it. If you truly want to ask a purely conceptual question, you should indicate this in your question.
```

Reviewer: I see what you mean. I added the `back and forth` and `guessing` part based off on what I often see happening in our discord. But they do indeed translate to the same thing, both causing unnecessary back and forth. I think your rewrite captures that nicely.

Reviewer 2: Hey @ManonLef. Did you see my version?

Reviewer: I did, but I want to make sure Rachel can deal with these in succession and can maintain the birdseye view. Since it seems you reviewed the same text as I did. 

I do feel confident about this addition I made that is not reflected in your review. I added that because this document is the one we link to in our discord on a daily basis when guiding people with their issues. 

But to not overcomplicate this thread, I will leave the ownership with Rachel and she can review your suggestion as well.

Reviewer 2: Ok gotcha 😁

> If your question does not provide any code or other relevant information, you will only receive theoretical and vague answers. This often causes frustration for everyone. You may become frustrated because the answers are not helpful. Those helping may become frustrated because as you add new information, the meaning of your original question changes and they wasted time trying to answer it. If you truly want to ask a purely conceptual question, you should indicate this in your question.

Reviewer 2: Interesting points @rlmoser99 and @ManonLef. Here's my version. I provided a little bit more detail about the emotions that helpers and learners tend to feel in this scenario so new students can fully grasp the importance of being mindful of the questions they post. I bolded the word "truly" for this reason as well. What do you think? 
```suggestion
If your question does not provide any code or other relevant information, you will only receive theoretical and vague answers. This often causes frustration for everyone. For you, the answers are not helping you solve the problem. Whereas for the helpers, the meaning of your original question changes due to you adding more context to it, causing them to feel they wasted time trying to answer your question. If you **truly** want to ask a purely conceptual question, you should indicate this in your question.
```

Author: Thanks for this suggestion. I see that using the word frustration/frustrated over again is repetitive, so I am going to work on a alternative. Also, I am not a big fan of the use of **bold** words.

> #### Ask About the Problem at Hand, Not the Solution Itself

Reviewer: ```suggestion
#### Ask about the problem at hand, not the solution itself
```

### Assignment

<div class="lesson-content__panel" markdown="1">
1.  [Don't ask to ask, just ask](https://dontasktoask.com/) demonstrates the importance of just asking your question and not asking to ask your question.
2.  Read about the [“XY Problem”](https://xyproblem.info/), which is a common pitfall both new and experienced programmers fall into when asking questions.
3.  Read and bookmark our community guide [How to Ask Technical Questions](https://www.theodinproject.com/guides/community/how_to_ask). It is a great reference to refer to whenever you find yourself needing to ask for help, and you might find yourself solving your own problem as you think about the points listed in the article.
</div>

> ### Knowledge Check

Reviewer: ```suggestion
### Knowledge check
```

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

*   [Describe the “XY Problem”.](https://xyproblem.info/)
*   [Name the 5 things your question should include.](https://www.theodinproject.com/guides/community/how_to_ask)

> ### Additional Resources

Reviewer: ```suggestion
### Additional resources
```
